MathlibTest/Junk.lean

```diff
@@ -0,0 +1,360 @@
+  norm_num
+
+/-! ## Division by zero -/
+
+/-- Dividing a `Nat` by zero gives zero. Similarly for `Int`. -/
+example : (1 : ℕ) / 0 = 0 ∧ (2 : ℤ) / 0 = 0 := by
+  norm_num
+
+/-! There's no division defined for `PNat`, so no worry about it giving 0 or 1. -/
+/--
+error: failed to synthesize
+  HDiv ℕ+ ℕ+ ℕ+
+
+Hint: Additional diagnostic information may be available using the `set_option diagnostics true` command.
+-/
+#guard_msgs in
+#synth HDiv ℕ+ ℕ+ ℕ+
+
+/-- In any `GroupWithZero`, division by zero gives the junk value zero. This includes, by extension,
+any `Field` (like `Rat` or `Real`), as well as types like `NNRat` and `NNReal`. -/
+example {G : Type*} [GroupWithZero G] :
+    (10 : NNRat) / 0 = 0 ∧
+    (20 : NNReal) / 0 = 0 ∧
+    ∀ (x : G), x / 0 = 0 := by
+  simp
+
+/-- Division by zero also gives 0 in some more exotic non-groups like `EReal` and `Ordinal`. -/
+example : (1 : EReal) / 0 = 0 ∧ Ordinal.omega0 / 0 = 0 := by
+  simp
+
+/-- Note that division by zero does _not_ give zero in `ENNReal`, where it gives `+∞` instead. -/
+example : (1 : ENNReal) / 0 = ⊤ := by
+  simp
+
+/-! ## Arithmetic with Fin -/
+
+/-- `Fin n` represents "natural numbers below n", and generally behaves like mod-n arithmetic. -/
+example :
+    (3 : Fin 5) + (4 : Fin 5) = 2 ∧
+    (3 : Fin 5) * (2 : Fin 5) = 1 ∧
+    (2 : Fin 9) ^ 4 = -2
+    := by
+  and_intros <;> rfl
+
+/-- The design of `Fin` is heavily motivated by using it define the types like `UInt32`, which
+describe twos-complement machine arithmetic. This means that division actually behaves like `Nat`
+division, and not modular division. In this example, `3` and `11` are inverses modulo 16 and so
+`(3 : Fin 16) * (11 : Fin 16) = 1`, but on the other hand `(1 : Fin 16) / (3 : Fin 16)` is
+defined to have the same value as the natural number division `1 / 3`, which rounds down to 0.
+-/
+example :
+    (3 : Fin 16) * (11 : Fin 16) = 1 ∧
+    (1 : Fin 16) / (3 : Fin 16) = 0 := by
+  and_intros <;> rfl
+
+/-- `Fin` also allows arbitrary numeric literals, and simply casts them mod n. This can lead to
+surprising looking true goals, like `x + 6 = x + 37`. Numeric literals can have a very
+context-dependent meaning. -/
+example (x : Fin 31) : x + 6 = x + 37 := by
+  simp
+
+/-! Here even the bare goal `6 = 37` turns out to be true. -/
+/--
+info: theorem _example.extracted_1 : 6 = 37 := sorry
+-/
+#guard_msgs in
+example {x : Fin 31} : x + 6 = x + 37 := by
+  rw [add_right_inj]; clear x
+  extract_goal
+  simp
+
+/-! ## Sqrt and Log-/
+
+/-- `Nat.sqrt` rounds down to the nearest integer, again, in order to provide a `Nat` output. -/
+example : Nat.sqrt 5 = 2 := by
+  norm_num
+
+/-- `Int.sqrt` and `Real.sqrt` obviously can take negative arguments, but we don't want to return
+imaginary numbers, so they return 0 for any negative arguments. -/
+example : (-3).sqrt = 0 ∧ (-Real.pi).sqrt = 0 := by
+  constructor
+  · rfl
+  · simp [Real.sqrt_eq_zero', Real.pi_nonneg]
```

Comment:
good place to showcase the "spec" lemma about `sqrt_sq`, if they exist. Int is maybe misnamed: `Int.sqrt_eq`, `Int.exists_mul_self`.